components/raftstore/src/store/peer_storage.rs

```diff
@@ -755,7 +795,137 @@ where
+                                if low <= last {
+                                    // the fetched entries covers [low, ..], high may be
+                                    let idx = (low - first) as usize;
+                                    return Ok(ents
+                                        .drain(
+                                            idx..std::cmp::min(
+                                                ents.len(),
+                                                idx + (high - low) as usize,
+                                            ),
+                                        )
+                                        .collect());
+                                } else {
+                                    // the fetched entries doesn't cover [low, high]
+                                    self.raftlog_fetch_stats.fetch_invalid.update(|m| m + 1);
+                                }
+                            } else if *first > low {
+                                info!(
+                                    "fallback to fetch the missing log entries [{}, {}) in sync way",
+                                    low, *first
+                                );
+                                self.raftlog_fetch_stats.fallback_fetch.update(|m| m + 1);
+                                let mut full_ents = vec![];
+                                self.engines.raft.fetch_entries_to(
+                                    region_id,
+                                    low,
+                                    *first,
+                                    Some(max_size as usize),
+                                    &mut full_ents,
+                                )?;
+                                if full_ents.last().map(|e| e.index).unwrap_or_default()
+                                    == *first - 1
+                                {
+                                    full_ents.append(&mut ents);
+                                };
+                                return Ok(full_ents);
+                            }
+                        }
+                    };
+                }
+                Err(e) => {
+                    if res.low == low {
+                        return Err(e);
+                    } else {
+                        self.raftlog_fetch_stats.fetch_invalid.update(|m| m + 1);
+                    }
+                }
+            }
+        }
+
+        self.raftlog_fetch_stats.async_fetch.update(|m| m + 1);
+        self.async_fetch_cache.borrow_mut().insert(to, None);
+        self.raftlog_fetch_scheduler
+            .schedule(RaftLogFetchTask::SendAppend {
+                region_id,
+                to_peer: to,
+                low,
+                high,
+                max_size: (max_size as usize),
+            })
+            .unwrap();
+        Err(raft::Error::Store(
+            raft::StorageError::LogTemporarilyUnavailable,
+        ))
+    }
+
+    pub fn entries(
```

Comment:
The current behavior of "entries()" if it was called twice with exactly same low, high, max_size, is that it will fetch the entry anyway.
But with this change, if it's called twice with same async_to parameter, it can call async_fetch which could return an LogTemporarilyUnavailable error. It changes the semantic of entries API.